### ISSUE TITLE
Metanorma >1.9.0 throws strange errors in PDF generation

### ISSUE THREAD

**Issue by lrosenthol** [metanorma/metanorma#357]
I upgraded to 1.9.3 - both locally on my Mac (via homebrew) and via Docker (metanorma/metanorma).  In both cases, I now receive the following errors during PDF productions.   This is from the *.pdf.err file.

```
WARNING: font '/Users/lrosenth/.fontist/fonts/Inter-VariableFont_slnt,wght.ttf' from the fontist manifest file is the variable font and doesn't supported by Apache FOP!
[FATAL] Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
Can't obtain table's widths information: javax.xml.transform.TransformerException: org.apache.fop.fo.ValidationException: Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
WARNING: Font file '/Users/lrosenth/.metanorma/fonts/SourceSansPro-Light.ttf' (font name 'Source Sans Pro', font style 'normal', font weight '300') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
WARNING: Font file '/Users/lrosenth/.metanorma/fonts/SourceSansPro-Regular.ttf' (font name 'Source Sans Pro', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
WARNING: Font file '/Users/lrosenth/.metanorma/fonts/SourceSansPro-Bold.ttf' (font name 'Source Sans Pro', font style 'normal', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Bold.ttf'.
WARNING: Font file '/Users/lrosenth/.metanorma/fonts/SourceSansPro-It.ttf' (font name 'Source Sans Pro', font style 'italic', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Italic.ttf'.
WARNING: Font file '/Users/lrosenth/.metanorma/fonts/SourceSansPro-BoldIt.ttf' (font name 'Source Sans Pro', font style 'italic', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-BoldItalic.ttf'.
WARNING: Font 'Helvetica' (font name 'Helvetica', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
WARNING: Font 'Helvetica' (font name 'Helvetica', font style 'normal', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Bold.ttf'.
WARNING: Font 'Helvetica' (font name 'Helvetica', font style 'italic', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Italic.ttf'.
WARNING: Font 'Helvetica' (font name 'Helvetica', font style 'italic', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-BoldItalic.ttf'.
WARNING: Font '"source code pro"' (font name '"source code pro"', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
WARNING: Font '"source code pro"' (font name '"source code pro"', font style 'normal', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Bold.ttf'.
WARNING: Font '"source code pro"' (font name '"source code pro"', font style 'italic', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Italic.ttf'.
WARNING: Font '"source code pro"' (font name '"source code pro"', font style 'italic', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-BoldItalic.ttf'.
WARNING: Font '"menlo"' (font name '"menlo"', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
WARNING: Font '"menlo"' (font name '"menlo"', font style 'normal', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Bold.ttf'.
WARNING: Font '"menlo"' (font name '"menlo"', font style 'italic', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Italic.ttf'.
WARNING: Font '"menlo"' (font name '"menlo"', font style 'italic', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-BoldItalic.ttf'.
WARNING: Font '"monaco"' (font name '"monaco"', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
WARNING: Font '"monaco"' (font name '"monaco"', font style 'normal', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Bold.ttf'.
WARNING: Font '"monaco"' (font name '"monaco"', font style 'italic', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Italic.ttf'.
WARNING: Font '"monaco"' (font name '"monaco"', font style 'italic', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-BoldItalic.ttf'.
WARNING: Font '"courier new"' (font name '"courier new"', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSansMono-Regular.ttf'.
WARNING: Font '"courier new"' (font name '"courier new"', font style 'normal', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSansMono-Bold.ttf'.
WARNING: Font '"courier new"' (font name '"courier new"', font style 'italic', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSansMono-Regular.ttf'.
WARNING: Font '"courier new"' (font name '"courier new"', font style 'italic', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSansMono-Bold.ttf'.
WARNING: Font 'monospace' (font name 'monospace', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
WARNING: Font 'monospace' (font name 'monospace', font style 'normal', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Bold.ttf'.
WARNING: Font 'monospace' (font name 'monospace', font style 'italic', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Italic.ttf'.
WARNING: Font 'monospace' (font name 'monospace', font style 'italic', font weight 'bold') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-BoldItalic.ttf'.

WARNING: error in the font '/Users/lrosenth/.metanorma/fonts/SourceSansPro-LightIt.ttf' registering.
Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
org.apache.fop.fo.ValidationException: Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
javax.xml.transform.TransformerException: org.apache.fop.fo.ValidationException: Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
	at org.apache.xalan.transformer.TransformerIdentityImpl.transform(TransformerIdentityImpl.java:502)
	at org.metanorma.fop.PDFGenerator.generateFOPIntermediateFormat(PDFGenerator.java:771)
	at org.metanorma.fop.PDFGenerator.runFOP(PDFGenerator.java:505)
	at org.metanorma.fop.PDFGenerator.convertmn2pdf(PDFGenerator.java:462)
	at org.metanorma.fop.PDFGenerator.process(PDFGenerator.java:299)
	at org.metanorma.fop.mn2pdf.main(mn2pdf.java:338)
Caused by: org.apache.fop.fo.ValidationException: Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
	at org.apache.fop.events.ValidationExceptionFactory.createException(ValidationExceptionFactory.java:38)
	at org.apache.fop.events.EventExceptionManager.throwException(EventExceptionManager.java:58)
	at org.apache.fop.events.DefaultEventBroadcaster$1.invoke(DefaultEventBroadcaster.java:173)
	at jdk.proxy2/jdk.proxy2.$Proxy9.idNotUnique(Unknown Source)
	at org.apache.fop.fo.FObj.checkId(FObj.java:240)
	at org.apache.fop.fo.FObj.startOfNode(FObj.java:188)
	at org.apache.fop.fo.flow.Block.startOfNode(Block.java:145)
	at org.apache.fop.fo.FOTreeBuilder$MainFOHandler.startElement(FOTreeBuilder.java:367)
	at org.apache.fop.fo.FOTreeBuilder$2.run(FOTreeBuilder.java:185)
	at org.apache.fop.fo.FOTreeBuilder$2.run(FOTreeBuilder.java:182)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:319)
	at org.apache.fop.fo.FOTreeBuilder.startElement(FOTreeBuilder.java:181)
	at org.apache.xalan.transformer.TransformerIdentityImpl.startElement(TransformerIdentityImpl.java:1073)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.startElement(AbstractSAXParser.java:518)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:376)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2726)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:114)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
	at org.apache.xalan.transformer.TransformerIdentityImpl.transform(TransformerIdentityImpl.java:485)
	... 5 more
---------
org.apache.fop.fo.ValidationException: Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
	at org.apache.fop.events.ValidationExceptionFactory.createException(ValidationExceptionFactory.java:38)
	at org.apache.fop.events.EventExceptionManager.throwException(EventExceptionManager.java:58)
	at org.apache.fop.events.DefaultEventBroadcaster$1.invoke(DefaultEventBroadcaster.java:173)
	at jdk.proxy2/jdk.proxy2.$Proxy9.idNotUnique(Unknown Source)
	at org.apache.fop.fo.FObj.checkId(FObj.java:240)
	at org.apache.fop.fo.FObj.startOfNode(FObj.java:188)
	at org.apache.fop.fo.flow.Block.startOfNode(Block.java:145)
	at org.apache.fop.fo.FOTreeBuilder$MainFOHandler.startElement(FOTreeBuilder.java:367)
	at org.apache.fop.fo.FOTreeBuilder$2.run(FOTreeBuilder.java:185)
	at org.apache.fop.fo.FOTreeBuilder$2.run(FOTreeBuilder.java:182)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:319)
	at org.apache.fop.fo.FOTreeBuilder.startElement(FOTreeBuilder.java:181)
	at org.apache.xalan.transformer.TransformerIdentityImpl.startElement(TransformerIdentityImpl.java:1073)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.startElement(AbstractSAXParser.java:518)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.scanStartElement(XMLNSDocumentScannerImpl.java:376)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl$FragmentContentDriver.next(XMLDocumentFragmentScannerImpl.java:2726)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLNSDocumentScannerImpl.next(XMLNSDocumentScannerImpl.java:114)
	at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
	at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
	at org.apache.xalan.transformer.TransformerIdentityImpl.transform(TransformerIdentityImpl.java:485)
	at org.metanorma.fop.PDFGenerator.generateFOPIntermediateFormat(PDFGenerator.java:771)
	at org.metanorma.fop.PDFGenerator.runFOP(PDFGenerator.java:505)
	at org.metanorma.fop.PDFGenerator.convertmn2pdf(PDFGenerator.java:462)
	at org.metanorma.fop.PDFGenerator.process(PDFGenerator.java:299)
	at org.metanorma.fop.mn2pdf.main(mn2pdf.java:338)
```

**Comment by lrosenthol** [comment]
Actually, this isn't new.  It appeared in 1.9.1, but apparently I was still on 1.9.0 in my local install.

I checked each docker image starting with 1.9.3 (then 1.9.2, 1.9.1) and finally 1.9.0 produced the output!

Happy to share any/all additional files that will help figure out what changed in 1.9.1 that broke me.

**Comment by Intelligent2013** [comment]
I've tested the PDF generation for the sample ISO document https://github.com/metanorma/mn-samples-iso/blob/main/sources/international-standard/rice-en.final.adoc via the latest docker metanorma image and the process ends successfully.

The .pdf.err log contains only non-critical warnings about the fonts:
```
WARNING: font '/config/fonts/Inter-VariableFont_slnt,wght.ttf' from the fontist manifest file is the variable font and doesn't supported by Apache FOP!
WARNING: Font file '/root/.metanorma/fonts/SourceSansPro-Light.ttf' (font name 'Source Sans Pro', font style 'normal', font weight '300') doesn't exist. Replaced by '/root/.metanorma/fonts/NotoSans-Regular.ttf'.
WARNING: Font file '/root/.metanorma/fonts/SourceSansPro-Regular.ttf' (font name 'Source Sans Pro', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/root/.metanorma/fonts/NotoSans-Regular.ttf'.
WARNING: Font file '/root/.metanorma/fonts/SourceSansPro-Bold.ttf' (font name 'Source Sans Pro', font style 'normal', font weight 'bold') doesn't exist. Replaced by '/root/.metanorma/fonts/NotoSans-Bold.ttf'.
WARNING: Font file '/root/.metanorma/fonts/SourceSansPro-It.ttf' (font name 'Source Sans Pro', font style 'italic', font weight 'normal') doesn't exist. Replaced by '/root/.metanorma/fonts/NotoSans-Italic.ttf'.
WARNING: Font file '/root/.metanorma/fonts/SourceSansPro-BoldIt.ttf' (font name 'Source Sans Pro', font style 'italic', font weight 'bold') doesn't exist. Replaced by '/root/.metanorma/fonts/NotoSans-BoldItalic.ttf'.
```
I'll investigate them.

@lrosenthol the warnings about these fonts:
```
WARNING: Font 'Helvetica' (font name 'Helvetica', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
...
WARNING: Font '"source code pro"' (font name '"source code pro"', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
.
WARNING: Font '"menlo"' (font name '"menlo"', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
...
WARNING: Font '"monaco"' (font name '"monaco"', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
...
WARNING: Font '"courier new"' (font name '"courier new"', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSansMono-Regular.ttf'.
...
WARNING: Font 'monospace' (font name 'monospace', font style 'normal', font weight 'normal') doesn't exist. Replaced by '/Users/lrosenth/.metanorma/fonts/NotoSans-Regular.ttf'.
...
```
looks like relate to the fonts which used in the SVG image (my assumption).

The critical error that blocks the PDF generation is
```
Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
org.apache.fop.fo.ValidationException: Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
javax.xml.transform.TransformerException: org.apache.fop.fo.ValidationException: Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
```
My assumption is  - ID "foreword" was used for a fre elements - one in the main document, another in the SVG image.
But it's very strange that the PDF generated successfully in the version 1.9.0...

It will be very helpful if you attach the source adoc or `...presentation.xml`. Thanks.

**Comment by lrosenthol** [comment]
Here is the presentation.xml, but I searched for "forward" and didn't find it (except in some normative text).  Going to check the SVG's - hadn't thought about that!   Let me know if you find something...

[ContentCredentials.presentation.xml.zip](https://github.com/metanorma/metanorma/files/14529855/ContentCredentials.presentation.xml.zip)

**Comment by lrosenthol** [comment]
I searched all the SVG files - no "forward' there either.

**Comment by ronaldtse** [comment]
Thanks @lrosenthol for reporting this. We’ve always knew that SVG fonts and its embedded content can trip compilation up but haven’t actually encountered such a file. 

Depending on the investigation outcome we may need to do something with regards to determining a set of “SVG requirements” and corresponding tooling because the world of SVG files is vast and profound…  (such as if it determines fonts, the user needs to decide whether to accept those fonts or override them).

**Comment by lrosenthol** [comment]
Ah - I see what you mean!   Something in that SVG file is problematic for the fonts.  No problem - that I can easily fix!

Really appreciate your tracking this down!

**Comment by lrosenthol** [comment]
I changed the fonts in that SVG, and it didn't fix the issue with current metanorma via Docker.   I am trying to "sanitize" the SVG and seeing if that helps...

**Comment by Intelligent2013** [comment]
> Here is an updated presentation.xml...
> [ContentCredentials.presentation.xml.zip](https://github.com/metanorma/metanorma/files/14532205/ContentCredentials.presentation.xml.zip)

@lrosenthol thank you for the updated XML! Now, I can repeat the error 
```xml
org.apache.fop.fo.ValidationException: Property ID "foreword" (found on "fo:block") previously used; ID values must be unique within a document! (See position 65:41)
```
It's bug in the ISO XSLT, occurs due the element `review` with `from="foreword" to="foreword"`:
```xml
<foreword id="foreword" obligation="informative" displayorder="2">
	<title>Foreword</title>
	...
	<review id="_e761ef6b-8203-d5a1-e38a-57a1cb66e1d1" reviewer="ISO" date="2017-01-01T00:00:00Z" type="todo" from="foreword" to="foreword">
```
ISO XSLT mix up `foreword` as the element's name (as in the example https://github.com/metanorma/mn-samples-iso/blob/main/sources/international-standard/sections-en/00-foreword.adoc?plain=1) vs. element's id (as in your instance).
I'll fix this issue.

As quick workaround solution, you can just remove the Foreword id `[[foreword]]` in your adoc.

**Comment by lrosenthol** [comment]
@Intelligent2013 That workaround works!  I can now build with the latest docker.  (though I now get a new cover page and an annotation about the Forward - but at least it builds)

**Comment by Intelligent2013** [comment]
@lrosenthol the critical error with `id="foreword"` fixed in the version 1.9.4. Now you can update to the latest version by `docker pull metanorma/metanorma` and restore id `[[foreword]]`.

**Comment by lrosenthol** [comment]
Fix confirmed on my end as well!
